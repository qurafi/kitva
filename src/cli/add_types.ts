import { readFile, writeFile } from "fs/promises";
import { resolve } from "path";

const type_defs = `import "kitva/ambient";`;

const locals_def = 'validation: import("kitva/presets/ajv/index").AppLocal';

export async function addTypes(cwd: string) {
	const app_d_ts = resolve(cwd, "src/app.d.ts");
	const content = await readFile(app_d_ts, "utf8");

	return writeFile(app_d_ts, editAppDts(content));
}

export function editAppDts(content: string) {
	let new_content = content;
	if (!content.includes(type_defs)) {
		new_content = `${type_defs}\n${content}`;
	}

	// this will be handled by a vite plugin to generate $types2 with all locals type setup to each route
	// if (!content.includes(locals_def)) {
	//     new_content = new_content.replace(
	//         /(\/\/)* (interface Locals {)/,
	//         (_, __, intrfc) => {
	//             return `${intrfc}\n\t${locals_def};\n\t`;
	//         }
	//     );
	// }

	return new_content;
}
