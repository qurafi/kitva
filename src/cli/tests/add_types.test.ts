import { expect, it } from "vitest";
import { editAppDts } from "../add_types.js";

const fixtures = {
    default: `// See https://kit.svelte.dev/docs/types#app
// for information about these interfaces
declare global {
    namespace App {
        // interface Error {}
        // interface Locals {}
        // interface PageData {}
        // interface Platform {}
    }
}

export {};`,
    with_locals: `// See https://kit.svelte.dev/docs/types#app
// for information about these interfaces
declare global {
    namespace App {
        // interface Error {}
        interface Locals {
            user: {name:string, id:string}
        }
        // interface PageData {}
        // interface Platform {}
    }
}

export {};`,
    with_locals_inline: `// See https://kit.svelte.dev/docs/types#app
// for information about these interfaces
declare global {
    namespace App {
        // interface Error {}
        interface Locals {user: {name:string, id:string}}
        // interface PageData {}
        // interface Platform {}
    }
}

export {};`,
    already_defined_inline: `// See https://kit.svelte.dev/docs/types#app
// for information about these interfaces
declare global {
    namespace App {
        // interface Error {}
        interface Locals {validation: import("kitva/presets/ajv/index").AppLocal;}
        // interface PageData {}
        // interface Platform {}
    }
}

export {};`,
    already_defined: `// See https://kit.svelte.dev/docs/types#app
// for information about these interfaces
declare global {
    namespace App {
        // interface Error {}
        interface Locals {
            validation: import("kitva/presets/ajv/index").AppLocal;
        }
        // interface PageData {}
        // interface Platform {}
    }
}

export {};`,
};

for (const [name, content] of Object.entries(fixtures)) {
    it(`should edit app.d.ts file correctly: ${name}`, () => {
        const result = editAppDts(content);
        expect(result).toMatchSnapshot();
    });
}
