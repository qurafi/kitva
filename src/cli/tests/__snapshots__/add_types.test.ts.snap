// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should edit app.d.ts file correctly: already_defined 1`] = `
"import \\"kitva/ambient\\";
// See https://kit.svelte.dev/docs/types#app
// for information about these interfaces
declare global {
    namespace App {
        // interface Error {}
        interface Locals {
            validation: import(\\"kitva/presets/ajv/index\\").AppLocal;
        }
        // interface PageData {}
        // interface Platform {}
    }
}

export {};"
`;

exports[`should edit app.d.ts file correctly: already_defined_inline 1`] = `
"import \\"kitva/ambient\\";
// See https://kit.svelte.dev/docs/types#app
// for information about these interfaces
declare global {
    namespace App {
        // interface Error {}
        interface Locals {validation: import(\\"kitva/presets/ajv/index\\").AppLocal;}
        // interface PageData {}
        // interface Platform {}
    }
}

export {};"
`;

exports[`should edit app.d.ts file correctly: default 1`] = `
"import \\"kitva/ambient\\";
// See https://kit.svelte.dev/docs/types#app
// for information about these interfaces
declare global {
    namespace App {
        // interface Error {}
        interface Locals {
	validation: import(\\"kitva/presets/ajv/index\\").AppLocal;
	}
        // interface PageData {}
        // interface Platform {}
    }
}

export {};"
`;

exports[`should edit app.d.ts file correctly: with_locals 1`] = `
"import \\"kitva/ambient\\";
// See https://kit.svelte.dev/docs/types#app
// for information about these interfaces
declare global {
    namespace App {
        // interface Error {}
       interface Locals {
	validation: import(\\"kitva/presets/ajv/index\\").AppLocal;
	
            user: {name:string, id:string}
        }
        // interface PageData {}
        // interface Platform {}
    }
}

export {};"
`;

exports[`should edit app.d.ts file correctly: with_locals_inline 1`] = `
"import \\"kitva/ambient\\";
// See https://kit.svelte.dev/docs/types#app
// for information about these interfaces
declare global {
    namespace App {
        // interface Error {}
       interface Locals {
	validation: import(\\"kitva/presets/ajv/index\\").AppLocal;
	user: {name:string, id:string}}
        // interface PageData {}
        // interface Platform {}
    }
}

export {};"
`;
